tools/hilow.wdl

```diff
@@ -0,0 +1,295 @@
+version 1.1
+
+task extract_promoters {
+    meta {
+        description: "Extract promoters from a GTF file"
+        help: "Find all transcripts, excluding the mitochondrial chromosome. Add a 2000bp buffer to the transcription start site to include promoter. Extract gene_id and transcript_id ($10 and $12). Subtract the suffix from the transcript ID. Convert to BED format"
+        outputs: {
+            promoter: "Promoter regions in BED format"
+        }
+    }
+
+       parameter_meta {
+        annotation: "GTF (optionally gzip compressed) file containing gene annotations"
+    }
+
+    input {
+        File annotation
+    }
+
+    String base = basename(annotation, ".gz")
+
+    #@ except: LineWidth
+    command <<<
+        set -euo pipefail
+
+        gunzip -c ~{annotation} > ~{base} \
+           || ln -sf ~{annotation} ~{base}
+
+
+        sed -i 's/ /\t/g' ~{base}
+
+        awk -F\\t '{ if($3 == "transcript" && $1 !~ "chrM")
+            if($7 =="+")
+                print $1 "++" $4-2000 "++" $4+2000 "++" substr($12,2,length($12)-5) "|" substr($10,2,length($10)-3)
+            else if($7=="-")
+                print $1 "++" $5-2000 "++" $5+2000 "++" substr($12,2,length($12)-5) "|" substr($10,2,length($10)-3) }' \
+            ~{base} \
+            | grep -v "unassigned" \
+            | sort \
+            | uniq \
+            | sed s/\+\+/\\t/g \
+            | sort -k1,1 -k2,2n > promoter_regions.bed
+    >>>
+
+    output {
+        File promoter = "promoter_regions.bed"
+    }
+
+    runtime {
+        cpu: 1
+        memory: "4 GB"
+        maxRetries: 1
+        container: "ubuntu:22.04"
```

Comment:
This is going to pull from DockerHub, won't it? That could cause rate limiting issues if a user is running many runs in parallel without a central docker cache. A rather niche edge case, but still we should use a different container registry imo.

Reply:
The docker docs include this blurb:

> When pulling Docker Verified Publisher images, attribution towards rate limiting is not applied. For more details, see [Docker Verified Publisher Program](https://docs.docker.com/docker-hub/repos/manage/trusted-content/dvp-program/).

I can't find any details, but I'm not sure if the limits apply to official images.

Reply:
O ok then. That seems fine. We can always update this if someone runs into issues.

Reply:
Do you have a preference on where to pull from? If we want to avoid docker hub, I suspect we'll have to pull and repush to GHCR.

Reply:
Looks like we commented at the same time. I'm fine with DockerHub given the blurb you shared.

Reply:
Unfortunately, I found additional comments on a GitHub issue on the Docker hub repo that indicates "Docker Official Images" are subject to limiting, even though "Docker Verified Publisher" images are not. That doesn't make any sense to me, but it is what it is. So we'll need to figure out an alternative.

Reply:
Ah crud. I'm sure there's a [near] equivalent image on Quay, right?

Reply:
I spent some time searching Quay. I am appalled at their search. it is truly awful. Using Google with `site:quay.io` is no more illuminating. There are some random looking images with `ubuntu` in the name, but no details and nothing official looking. There is a mirror of the Docker official images on Amazon's ECR, but it also limits pulls without credentials.